### ISSUE TITLE
OSError: [Errno 63] File name too long

### ISSUE THREAD

**Issue by obruchez** [s3tools/s3cmd#202]
I'm trying to archive my old Jungle Disk backups from S3 and I get an error caused by a long filename (see below).

This happens with version 1.5.0-alpha3 of s3cmd, on MacOS X 10.8.4.

---

s3://bd26d51143aaa5e068257ac69c5e5717-default/13.backups.AGHARTA.C.Documents and Settings.Aline.My Documents.Uni.1_c3_a8re ann_c3_a9e.Histoire.Histoire g_c3_a9n_c3_a9rale.S_c3_a9minaire Mariage.docs en vrac.Le mariage dans le pays de Vaud _c3_a0 l_e2_80_99_c3_a9poque des Lumi_c3_a8res_c2_a0(plan)_-doc.file -> <fdopen>  [13667 of 19718]
 72704 of 72704   100% in    0s   349.02 kB/s  done

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    An unexpected error has occurred.
  Please report the following lines to:
   s3tools-bugs@lists.sourceforge.net
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Problem: OSError: [Errno 63] File name too long
S3cmd:   1.5.0-alpha3

Traceback (most recent call last):
  File "/usr/local/bin/s3cmd", line 2094, in <module>
    main()
  File "/usr/local/bin/s3cmd", line 2034, in main
    cmd_func(args)
  File "/usr/local/bin/s3cmd", line 1193, in cmd_sync
    return cmd_sync_remote2local(args)
  File "/usr/local/bin/s3cmd", line 891, in cmd_sync_remote2local
    seq, total_size = _download(remote_list, seq, remote_count + update_count, total_size, dir_cache)
  File "/usr/local/bin/s3cmd", line 850, in _download
    raise e
OSError: [Errno 63] File name too long

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    An unexpected error has occurred.
    Please report the above lines to:
   s3tools-bugs@lists.sourceforge.net
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by m13** [comment]
+1

**Comment by fviard** [comment]
Hi,
As the error is telling you, these file names are too long to be supported by your filesystem of the system you are downloading to.
Most filesystems today have a limit of 255 characters for the file name.
For @obruchez case, the file name does: 259 == 255+".file" . So I suppose that the software that you used to backup to s3 did add a final ".file" string to the filename that did exceed the 255 char limit.

For @m13, the filename looks to be way too long exceeding the 255 char limit.
This is probably because you have a lot of special characters in it.

So, it is nothing related to s3cmd and so there is nothing that s3cmd can do about it.
The suggestions that I could have to help you:
- Rename the problematic files on s3 before trying to download them:
s3cmd mv s3://xxxxxx/very-long-a-b-c...txt s3://xxxxxxx/abc
- Format an usb drive with a filesystem tweaked to have a file name limit higher than 255 and try to download the data inside it instead of inside you computer volume.

**Comment by m13** [comment]
Thanks @fviard !
Maybe s3cmd can handle the exception and let you continue with the next file if you set "--ignore-failed-copy" or similar.

**Comment by obruchez** [comment]
Another idea might be for s3cmd to detect that the object name is too long, automatically choose a shorter filename and save the mapping somewhere. Anyway, I finally used Amazon's official CLI tools to download my backup bucket.

**Comment by mdomsch** [comment]
I don't want to be in the business of doing longname to shortname mappings
(ala FAT32).  I'd much rather WARN or ERROR out on such and let the caller
deal with it.

**Comment by fviard** [comment]
@obruchez I'm curious, when you did download your backup bucket with Amazon cli, did you see the file that was causing the error with S3cmd in your destination folder? if yes, was it renamed, or had his name truncated?

**Comment by m13** [comment]
@fviard I did the same and you don't have those files in your backup, but they are reported in the console output so you can manage them later (for example, "$ s3cmd get" works).

**Comment by obruchez** [comment]
@fviard The problematic files were indeed skipped.

**Comment by slorber** [comment]
I have 60go to transfert and the download stops everytime this happens.
I can use the de/rename command on these files and try to sync again but this stops for every long filename so it's quite annoying. Hopefully I won't have much of these files.

**Comment by fviard** [comment]
@m13 @obruchez @slorber 
I did a fix for this issue to have the "skip" working correctly, can you give a try to the following branch:
https://github.com/fviard/s3cmd/tree/upload_download_errors_rework
Thanks

**Comment by slorber** [comment]
@fviard unfortunatly I don't DL files from S3 but from GreenQloud (compatible REST API with S3) and I was not able to run master code of s3cmd on it (but it works with 1.1.0)  so I won't be able to test...

**Comment by mdomsch** [comment]
Why is master branch not working?  Use --signature-v2 command line option
to force older signature method...

**Comment by fviard** [comment]
@slorber That is unfortunate.
I think that I know why GreenQloud, if it is what I think, it doesn't support yet batch delete.
But so I don't see why download (/remote2local) would not work.
Could it be possible for you to try on a small subset/subpath with the "-d" option and to give us the debug log?

**Comment by slorber** [comment]
@mdomsch I can confirm this works fine with `--signature-v2`. Actually I'm not an S3 expert and tested before knowing anything regarding to V2 vs V4. As GreenQloud did suggest in their migration documentation to use 1.0 I did not try a lot with master, but yes it works fine (at least with the `du` command I don't get `AuthMethodNotRecognized` anymore)

@fviard I'll try to test your commit but I'm not sure to be able to do so, as I deleted the problematic files and I'm not sure there will be additional ones.
